go-controller/pkg/ovn/policy.go

```diff
@@ -51,6 +48,90 @@ const (
 	arpAllowPolicySuffix = "ARPallowPolicy"
 )
 
+// defaultDenyPortGroups is a shared object and should be used by only 1 thread at a time
+type defaultDenyPortGroups struct {
+	// countersMutex is needed to sync portCounters access
+	countersMutex sync.Mutex
+	// portName: map[portName]int
+	// count policies that are using every port in the map
+	// these counter should be atomically updated with db operations
+	// if adding a port to db fails, counter shouldn'd be changed
+	ingressPortCounters map[string]int
+	egressPortCounters  map[string]int
+	// policies is a map of policies that use this port group
+	// policy keys must be unique, and it can be retrieved with (np *networkPolicy) getKey()
+	policies map[string]bool
+}
+
+// updateDefaultDenyPortsCountersAdd increments counters for default deny port group ports.
+// If port should be added to ingress and/or egress default deny port group depends on policy spec.
+func (sharedPGs *defaultDenyPortGroups) updateDefaultDenyPortsCountersAdd(np *networkPolicy,
+	portNamesToUUIDs map[string]string) (ingressDenyPorts, egressDenyPorts []string) {
+	ingressDenyPorts = []string{}
+	egressDenyPorts = []string{}
+
+	sharedPGs.countersMutex.Lock()
+	defer sharedPGs.countersMutex.Unlock()
+
+	if np.isIngress {
+		for portName, portUUID := range portNamesToUUIDs {
+			// if this is the first NP referencing this pod, then we
+			// need to add it to the port group.
+			if sharedPGs.ingressPortCounters[portName] == 0 {
+				ingressDenyPorts = append(ingressDenyPorts, portUUID)
+			}
+			// increment the reference count.
+			sharedPGs.ingressPortCounters[portName]++
+		}
+	}
+	if np.isEgress {
+		for portName, portUUID := range portNamesToUUIDs {
+			if sharedPGs.egressPortCounters[portName] == 0 {
+				// again, reference count is 0, so add to port
+				egressDenyPorts = append(egressDenyPorts, portUUID)
+			}
+			// bump reference count
+			sharedPGs.egressPortCounters[portName]++
+		}
+	}
+	return
+}
+
+// updateDefaultDenyPortsCountersDel decrements counters for default deny port group ports, and deletes port
+// from the default-deny portgroup if the reference count for the pod is 0
+func (sharedPGs *defaultDenyPortGroups) updateDefaultDenyPortsCountersDel(np *networkPolicy,
+	portNamesToUUIDs map[string]string) (ingressDenyPorts, egressDenyPorts []string) {
+	ingressDenyPorts = []string{}
+	egressDenyPorts = []string{}
+
+	sharedPGs.countersMutex.Lock()
+	defer sharedPGs.countersMutex.Unlock()
+
+	if np.isIngress {
+		for portName, portUUID := range portNamesToUUIDs {
+			if sharedPGs.ingressPortCounters[portName] > 0 {
+				sharedPGs.ingressPortCounters[portName]--
+				if sharedPGs.ingressPortCounters[portName] == 0 {
```

Comment:
Yeah this will be an implementation error, and will mean that port was deleted from default deny port group, but not from localPods (which should never happen), but I guess in that case just deleting already deleted port should be ok